runtime/src/iree/hal/drivers/amdgpu/util/libhsa.c

```diff
@@ -0,0 +1,692 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/hal/drivers/amdgpu/util/libhsa.h"
+
+#include "iree/base/internal/debugging.h"
+#include "iree/base/internal/dynamic_library.h"
+#include "iree/base/internal/path.h"
+
+//===----------------------------------------------------------------------===//
+// hsa_status_t interop
+//===----------------------------------------------------------------------===//
+
+// Maps an HSA status to an IREE status as best we can.
+static iree_status_code_t iree_hsa_status_code(hsa_status_t status) {
+  switch (status) {
+    default:
+      return IREE_STATUS_UNKNOWN;
+    case HSA_STATUS_SUCCESS:
+      return IREE_STATUS_OK;
+    case HSA_STATUS_INFO_BREAK:
+      return IREE_STATUS_CANCELLED;
+    case HSA_STATUS_ERROR:
+      return IREE_STATUS_UNKNOWN;
+    case HSA_STATUS_ERROR_INVALID_ARGUMENT:
+    case HSA_STATUS_ERROR_INVALID_QUEUE_CREATION:
+    case HSA_STATUS_ERROR_INVALID_ALLOCATION:
+    case HSA_STATUS_ERROR_INVALID_AGENT:
+    case HSA_STATUS_ERROR_INVALID_REGION:
+    case HSA_STATUS_ERROR_INVALID_SIGNAL:
+    case HSA_STATUS_ERROR_INVALID_QUEUE:
+    case HSA_STATUS_ERROR_INVALID_PACKET_FORMAT:
+    case HSA_STATUS_ERROR_INVALID_ISA_NAME:
+    case HSA_STATUS_ERROR_INVALID_CODE_OBJECT:
+    case HSA_STATUS_ERROR_INVALID_EXECUTABLE:
+    case HSA_STATUS_ERROR_INVALID_CODE_SYMBOL:
+    case HSA_STATUS_ERROR_INVALID_EXECUTABLE_SYMBOL:
+    case HSA_STATUS_ERROR_INVALID_FILE:
+    case HSA_STATUS_ERROR_INVALID_CODE_OBJECT_READER:
+    case HSA_STATUS_ERROR_INVALID_CACHE:
+    case HSA_STATUS_ERROR_INVALID_WAVEFRONT:
+    case HSA_STATUS_ERROR_INVALID_SIGNAL_GROUP:
+    case HSA_STATUS_ERROR_INVALID_RUNTIME_STATE:
+      return IREE_STATUS_INVALID_ARGUMENT;
+    case HSA_STATUS_ERROR_RESOURCE_FREE:
+      return IREE_STATUS_INTERNAL;
+    case HSA_STATUS_ERROR_INCOMPATIBLE_ARGUMENTS:
+    case HSA_STATUS_ERROR_INVALID_ISA:
+      return IREE_STATUS_INCOMPATIBLE;
+    case HSA_STATUS_ERROR_OUT_OF_RESOURCES:
+    case HSA_STATUS_ERROR_REFCOUNT_OVERFLOW:
+      return IREE_STATUS_RESOURCE_EXHAUSTED;
+    case HSA_STATUS_ERROR_NOT_INITIALIZED:
+    case HSA_STATUS_ERROR_FROZEN_EXECUTABLE:
+    case HSA_STATUS_ERROR_VARIABLE_ALREADY_DEFINED:
+    case HSA_STATUS_ERROR_VARIABLE_UNDEFINED:
+      return IREE_STATUS_FAILED_PRECONDITION;
+    case HSA_STATUS_ERROR_INVALID_INDEX:
+      return IREE_STATUS_OUT_OF_RANGE;
+    case HSA_STATUS_ERROR_INVALID_SYMBOL_NAME:
+      return IREE_STATUS_NOT_FOUND;
+    case HSA_STATUS_ERROR_EXCEPTION:
+    case HSA_STATUS_ERROR_FATAL:
+      return IREE_STATUS_DATA_LOSS;
+  }
+}
+
+// Returns the stringified form of the HSA status enum value.
+static const char* iree_hsa_status_string(hsa_status_t status) {
+  switch (status) {
+    default:
+      return "?";
+    case HSA_STATUS_SUCCESS:
+      return "HSA_STATUS_SUCCESS";
+    case HSA_STATUS_INFO_BREAK:
+      return "HSA_STATUS_INFO_BREAK";
```

Comment:
Do we have a status define that could guard this code? I'm wondering if we want a minimal binary size configuration that doesn't include all of these extra strings, here and down in `iree_hsa_status_description()`

Reply:
So long as the only code calling it is our status constructors it should be DCEd - `iree_status_from_hsa_status` -> `iree_make_status_with_location(these methods)` and if we aren't capturing messages (code-only) then `iree_make_status_with_location` becomes `IREE_STATUS_IMPL_MAKE_(code)` and all the other args are ignored.

(but good looking out :)